sympy/polys/numberfields/tests/test_subfield.py

```diff
@@ -271,6 +272,8 @@ def test_primitive_element():
     a, b = I*sqrt(2*sqrt(2) + 3), I*sqrt(-2*sqrt(2) + 3)
     assert primitive_element([a, b, I], x) == (x**4 + 6*x**2 + 1, [1, 0, 0])
 
+    assert primitive_element([sqrt(2), 0], x, ex=True) == (x**2 - 2, [1, 0], [[MPQ(1,1), MPQ(0,1)], []])
```

Comment:
Without `ex=True` this still gives an error:
```python
In [3]: primitive_element([sqrt(2), 0])
---------------------------------------------------------------------------
OptionError
```

Reply:
Should'nt that give an error? or should it's output be same as that of with `ex=True`.

Reply:
Might want to also have a test case where `ex=True` but the _first_ extension is rational, such as

```python
primitive_element([0, sqrt(2)], x, ex=True)
```

since I'm wondering if in such a case these lines

https://github.com/sympy/sympy/blob/403067f7bd22d63693a152c659d8681a3125fb9b/sympy/polys/numberfields/subfield.py#L364-L367

might cause trouble. Maybe here if `gen.isRational` then `K = QQ` and adjust `coeffs` and `reps` accordingly (and don't compute `f`).

Reply:
> ```python
> primitive_element([0, sqrt(2)], x, ex=True)
> ```

Yeah this test is failing with this error,
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/praneeth/sympy/sympy/polys/numberfields/subfield.py", line 387, in primitive_element
    H = [_.rep for _ in reps]
  File "/home/praneeth/sympy/sympy/polys/numberfields/subfield.py", line 387, in <listcomp>
    H = [_.rep for _ in reps]
AttributeError: 'PythonMPQ' object has no attribute 'rep'
```

Reply:
> @praneethratna I don't think it should be an error. It's just that when `ex=False` then it's a completely separate block of code,

Oh, then seems like this is not the correct fix then. I'm not sure of how this can be fixed.

Reply:
Maybe the whole function needs a rewrite.

The following is passing all tests under `sympy/polys`. I have not tested it outside of there, though.
(Note: need to import `_sympify` at top of file.)

```python
    if not extension:
        raise ValueError("Cannot compute primitive element for empty extension")
    extension = [_sympify(ext) for ext in extension]

    if x is not None:
        x, cls = sympify(x), Poly
    else:
        x, cls = Dummy('x'), PurePoly

    coeffs = []
    reps = []
    # Seek first irrational generator
    for i, gen in enumerate(extension):
        if gen.is_Rational:
            coeffs.append(0)
            reps.append([gen] if gen else [])
        else:
            coeffs.append(1)
            f = minimal_polynomial(gen, x, polys=True)
            if ex:
                K = QQ.algebraic_field((f, gen))
                reps.append(K.unit.rep)
            break
    else:
        # There were no irrational generators
        f = Poly(x)

    if not ex:
        for ext in extension[i+1:]:
            if ext.is_Rational:
                coeffs.append(0)
                continue
            _, factors = factor_list(f, extension=ext)
            f = _choose_factor(factors, x, gen)
            s, _, f = f.sqf_norm()
            gen += s*ext
            coeffs.append(s)

        if not polys:
            return f.as_expr(), coeffs
        else:
            return cls(f), coeffs

    for ext in extension[i+1:]:
        if ext.is_Rational:
            coeffs.append(0)    # rational ext is not included in the expression of a primitive element
            reps.append(K.convert(ext).rep)    # but it is included in reps
            continue
        p = minimal_polynomial(ext, x, polys=True)
        L = QQ.algebraic_field((p, ext))
        _, factors = factor_list(f, domain=L)
        f = _choose_factor(factors, x, gen)
        s, g, f = f.sqf_norm()
        gen += s*ext
        coeffs.append(s)
        K = QQ.algebraic_field((f, gen))
        h = _switch_domain(g, K)
        erep = _linsolve(h.gcd(p))  # ext as element of K
        ogen = K.unit - s*erep  # old gen as element of K
        reps = [dup_eval(_, ogen, K).rep for _ in reps] + [erep.rep]

    if not polys:
        return f.as_expr(), coeffs, reps
    else:
        return f, coeffs, reps
```

Reply:
I would keep the two ex cases separate and only add these lines to the False case:

    If ext.is_Rational:
        coeffs.append(0)
        continue

Reply:
> I would keep the two ex cases separate and only add these lines to the False case:

There are few failing test cases, like `primitive_element([sqrt(2), sqrt(3)], x)` gives `(x**2 - 2, [1])` which is incorrect i guess.

Reply:
This is what the current implementation gives:
```
In [11]: primitive_element([sqrt(2), sqrt(3)], x)                               
Out[11]: 
⎛ 4       2            ⎞
⎝x  - 10⋅x  + 1, [1, 1]⎠
```
Do you mean that adding the three lines in https://github.com/sympy/sympy/pull/23099#discussion_r810463900 would change the result?

Reply:
No, what is meant is that this implementation for `ex=False` case, the code is
```python
if not ex:
        for ext in extension[i+1:]:
            if ext.is_Rational:
                coeffs.append(0)
                continue
            _, factors = factor_list(f, extension=ext)
            f = _choose_factor(factors, x, gen)
            s, _, f = f.sqf_norm()
            gen += s*ext
            coeffs.append(s)
```
But If I'm correct what you have suggested is
```python
if not ex:
        for ext in extension[i+1:]:
            if ext.is_Rational:
                coeffs.append(0)
                continue
```
then the output is changing.

Reply:
The current implementation looks like this:
```
    if not ex:
        gen, coeffs = extension[0], [1]
        g = minimal_polynomial(gen, x, polys=True)
        for ext in extension[1:]:
            _, factors = factor_list(g, extension=ext)
            g = _choose_factor(factors, x, gen)
            s, _, g = g.sqf_norm()
            gen += s*ext
            coeffs.append(s)
```
After this change:
```
    if not ex:
        gen, coeffs = extension[0], [1]
        g = minimal_polynomial(gen, x, polys=True)
        for ext in extension[1:]:
            if ext.is_Rational:
                coeffs.append(0)
                continue
            _, factors = factor_list(g, extension=ext)
            g = _choose_factor(factors, x, gen)
            s, _, g = g.sqf_norm()
            gen += s*ext
            coeffs.append(s)
```
I get
```
In [1]: primitive_element([sqrt(2), sqrt(3)], x)                                
Out[1]: 
⎛ 4       2            ⎞
⎝x  - 10⋅x  + 1, [1, 1]⎠
```

Reply:
Sorry, i thought the implementation that you meant was https://github.com/sympy/sympy/pull/23099#discussion_r810397651, Yeah the change that you have mentioned is working except for one case `primitive_element([0, sqrt(2)], x, ex=True)` which is still giving an error.

Reply:
It seems that `dup_eval` returns an mpq object. I don't think that should happen.

Reply:
I'm not sure but this change is fixing the error, But some test cases are failing.
```diff
diff --git a/sympy/polys/numberfields/subfield.py b/sympy/polys/numberfields/subfield.py
index d9e3aec008..343b795347 100644
--- a/sympy/polys/numberfields/subfield.py
+++ b/sympy/polys/numberfields/subfield.py
@@ -382,13 +382,12 @@ def primitive_element(extension, x=None, *, ex=False, polys=False):
         h = _switch_domain(g, K)
         erep = _linsolve(h.gcd(p))  # ext as element of K
         ogen = K.unit - s*erep  # old gen as element of K
-        reps = [dup_eval(_.rep, ogen, K) for _ in reps] + [erep]
+        reps = [dup_eval(_, ogen, K).rep for _ in reps] + [erep.rep]

-    H = [_.rep for _ in reps]
     if not polys:
-        return f.as_expr(), coeffs, H
+        return f.as_expr(), coeffs, reps
     else:
-        return f, coeffs, H
+        return f, coeffs, reps

```

Reply:
This will make `dup_eval` return an ANP object with `rep`.
```
--- a/sympy/polys/densetools.py
+++ b/sympy/polys/densetools.py
@@ -273,9 +273,6 @@ def dup_eval(f, a, K):
     11
 
     """
-    if not a:
-        return dup_TC(f, K)
-
     result = K.zero
 
     for c in f:
```
It will give
```
In [1]: primitive_element([0, sqrt(2)], x, ex=True)                             
Out[1]: (x**2 - 2, [1, 1], [[], [mpq(1,1), mpq(0,1)]])
```

Reply:
Yeah, every test case is passing now, I'll push these changes.

Reply:
This seems to work as well:
```
--- a/sympy/polys/densetools.py
+++ b/sympy/polys/densetools.py
@@ -274,7 +274,7 @@ def dup_eval(f, a, K):
 
     """
     if not a:
-        return dup_TC(f, K)
+        return K.convert(dup_TC(f, K))
 
     result = K.zero
```
That might be slightly more efficient though the case `a == 0` is probably not common.

Reply:
So should i commit this change or the previous change is fine?

Reply:
This change goes straight to the essential point, so maybe it would be preferable.